### PR TITLE
gnrc: logging

### LINKED ISSUE #3502: log: use logging instead of DEBUG where appropriate for warnings and errors
Some error messages and warnings in gnrc (and probably elsewhere) are only printed if DEBUG for this module is enabled. While it is okay to discard error messages for production code, it would make testing and debugging a lot more easy if these warnings and errors would always show up, if compiled with `DEVELHELP`. Using `LOG_WARNING` or `LOG_ERROR` in these cases would be the better choice.

### PR DESCRIPTION
**Pull request by OlegHahm**
Depends on #3505.

Fixes #3502.

### CODE REVIEW

File: sys/net/link_layer/ng_netdev_eth/ng_netdev_eth.c [hunk 1]
```diff
@@ -99,37 +97,37 @@ static int _send_data(ng_netdev_t *netdev, ng_pktsnip_t *pkt)
...
-    DEBUG("ng_netdev_eth: send data ");
+    LOG_DEBUG("ng_netdev_eth: send data ");
 
     if (pkt == NULL) {
         return -EFAULT;
     }
 
     if ((dev == NULL) || (netdev->driver != &ng_netdev_eth_driver)) {
-        DEBUG("[wrong device descriptor]\n");
+        LOG_ERROR("[wrong device descriptor]\n");
```
Comment (authmillenon) [inline sys/net/link_layer/ng_netdev_eth/ng_netdev_eth.c]: Line is started as debug above.

File: sys/net/link_layer/ng_netdev_eth/ng_netdev_eth.c [hunk 2]
```diff
@@ -99,37 +97,37 @@ static int _send_data(ng_netdev_t *netdev, ng_pktsnip_t *pkt)
...
-    DEBUG("\n");
+    LOG_DEBUG("\n");
 
     to_send = _marshall_ethernet(dev, send_buffer, pkt);
     ng_pktbuf_release(pkt);
 
     if (to_send < 0) {
         errno = -to_send;
-        DEBUG("marshall\n");
+        LOG_ERROR("marshall\n");
```
Comment (authmillenon) [inline sys/net/link_layer/ng_netdev_eth/ng_netdev_eth.c]: Same here

File: sys/net/link_layer/ng_netdev_eth/ng_netdev_eth.c [hunk 3]
```diff
@@ -138,19 +136,19 @@ static int _send_data(ng_netdev_t *netdev, ng_pktsnip_t *pkt)
 
 static int _add_event_callback(ng_netdev_t *dev, ng_netdev_event_cb_t cb)
 {
-    DEBUG("ng_netdev_eth: add event callback");
+    LOG_DEBUG("ng_netdev_eth: add event callback");
 
     if ((dev == NULL) || (dev->driver != &ng_netdev_eth_driver)) {
-        DEBUG(" [wrong device descriptor]\n");
+        LOG_ERROR(" [wrong device descriptor]\n");
```
Comment (authmillenon) [inline sys/net/link_layer/ng_netdev_eth/ng_netdev_eth.c]: And here

File: sys/net/link_layer/ng_netdev_eth/ng_netdev_eth.c [hunk 4]
```diff
@@ -138,19 +136,19 @@ static int _send_data(ng_netdev_t *netdev, ng_pktsnip_t *pkt)
...
         return -ENODEV;
     }
 
     if (dev->event_cb != NULL) {
-        DEBUG(" [no space left]\n");
+        LOG_ERROR(" [no space left]\n");
```
Comment (authmillenon) [inline sys/net/link_layer/ng_netdev_eth/ng_netdev_eth.c]: And here

File: sys/net/link_layer/ng_netdev_eth/ng_netdev_eth.c [hunk 5]
```diff
@@ -158,19 +156,19 @@ static int _add_event_callback(ng_netdev_t *dev, ng_netdev_event_cb_t cb)
 
 static int _rem_event_callback(ng_netdev_t *dev, ng_netdev_event_cb_t cb)
 {
-    DEBUG("ng_netdev_eth: remove event callback");
+    LOG_DEBUG("ng_netdev_eth: remove event callback");
 
     if ((dev == NULL) || (dev->driver != &ng_netdev_eth_driver)) {
-        DEBUG(" [wrong device descriptor]\n");
+        LOG_ERROR(" [wrong device descriptor]\n");
```
Comment (authmillenon) [inline sys/net/link_layer/ng_netdev_eth/ng_netdev_eth.c]: and so on.

File: sys/net/application_layer/ng_zep/ng_zep.c [hunk 6]
```diff
@@ -34,10 +35,7 @@
 
 #include "net/ng_zep.h"
 
-#define ENABLE_DEBUG    (0)
-#include "debug.h"
-
-#if ENABLE_DEBUG
+#if (LOG_LEVEL > LOG_LEVEL_KERNEL)
 /* For PRIu16 etc. */
 #include <inttypes.h>
```
Comment (kaspar030) [inline sys/net/application_layer/ng_zep/ng_zep.c]: As we might have those line all over the place, we could move the include to log.h...

Comment (OlegHahm) [inline sys/net/application_layer/ng_zep/ng_zep.c]: Good idea.

File: sys/net/link_layer/ng_netdev_eth/ng_netdev_eth.c (hunk 1 above)
Comment (OlegHahm) [inline sys/net/link_layer/ng_netdev_eth/ng_netdev_eth.c]: I saw this, but would you propose?

Comment (authmillenon) [inline sys/net/link_layer/ng_netdev_eth/ng_netdev_eth.c]: Either making it too distinct messages or making one of them the level of the other.

**Comment by kaspar030** [comment]
Somehow I don't like this. It goes into the wrong direction.

```DEBUG``` was off by default. I agree that errors shouldn't be silently ignored, but now half of the messages have a log level more verbose than LOG_DEBUG. 

Remember having messages as LOG_ERROR will make them more likely to stay in the code, even if they are barely used.

Also, e.g. zep has this:
```"zep: could not allocate ZEP header in pktbuf\n"```
```"zep: could not allocate IPV6 header in pktbuf\n"```
```"zep: could not allocate UDP header in pktbuf\n"```
Replacing those with "zep: OOM\n" will probably save 150b of code, more could be saved when using a shared error condition.

Obvious bugs (unintended state) should use assert. Nothing should end up in any production code as the bug would be fixed.

We need to be careful that we don't go from "no error messages unless DEBUG is enabled", to "bloated code unless compiled silent".

Maybe a simple build size comparison would convince me otherwise, but I doubt it. (Just tried, ng_networking on sam-r21 grows 20% from master to this PR with default settings.)

**Comment by kaspar030** [comment]
In other words, this PR changes the log policy of a lot of debug statements, leading to log bloat.

File: sys/net/link_layer/ng_netdev_eth/ng_netdev_eth.c (hunk 1 above)
Comment (kaspar030) [inline sys/net/link_layer/ng_netdev_eth/ng_netdev_eth.c]: > I saw this, but would you propose?

assert()

**Comment by OlegHahm** [comment]
I partly agree. However, I would still argue that this goes into the right direction: making debugging more convenient.

The goal should be that we have the same code size (or less) after switching to this log architectures all over the place compared to the old `DEBUG` solution. In order to achieve this, `LOG_LEVEL_KERNEL` (or `LOG_LEVEL_NONE` or similar) should be a sensible default. With `DEVELHELP` I want to see at least all error messages (maybe also warnings). There's certainly room for improvement, for assigning the proper severity levels. I just assigned what seemed appropriate at a first glance. Having a little bit more code usage with `DEVELHELP`, but saving hours for searching a bug that doesn't exist seems a good trade-off to me.

**Comment by kaspar030** [comment]
> The goal should be that we have the same code size (or less) after switching to this log architectures all over the place compared to the old DEBUG solution. In order to achieve this, LOG_LEVEL_KERNEL (or LOG_LEVEL_NONE or similar) should be a sensible default.

That's cheating. The goal should be same or smaller codesize as with the old DEBUGing, with the same log messages, but room for finer grained improvements.

IMHO any DEBUG statement converted to more than LOG_DEBUG should be discussed seperately.

Remember, "zep: no memory in pktbuf for UDP header" costs 40b plus function call overhead in .text. I usually spent a lot of time to save that space.

**Comment by OlegHahm** [comment]
> That's cheating. The goal should be same or smaller codesize as with the old DEBUGing, with the same log messages, but room for finer grained improvements.

I don't see why. I told you from the beginning that I don't think that `LOG_INFO` is a sensible default value.

> IMHO any DEBUG statement converted to more than LOG_DEBUG should be discussed seperately.

That's what I did in #3502.

> Remember, "zep: no memory in pktbuf for UDP header" costs 40b plus function call overhead in .text. I usually spent a lot of time to save that space.

I don't argue that this can be converted into `LOG_DEBUG`. Just let us make sure that everything that reveals a critical issue, gets printed with `DEVELHELP`.

**Comment by kaspar030** [comment]
> I don't see why.

I just saw my application grow 20% in size.

> I told you from the beginning that I don't think that LOG_INFO is a sensible default value.

Possible. I also get more and more the feeling that integer log levels just don't cut it.

> That's what I did in #3502.

Not all of them at once, *every single*.

> I don't argue that this can be converted into LOG_DEBUG. 

*all* DEBUG statements should be converted to LOG_DEBUG. Off by default. No change in policy. No radical change in code size for the default compiles. We can improve later.

> Just let us make sure that everything that reveals a critical issue, gets printed with DEVELHELP.

Don't you think that raising a log message's severity level is the wrong way? They're not even linked to DEVELHELP...

I think we should draft up and discuss the concept of "logging" in a more general way.

**Comment by OlegHahm** [comment]
> >  That's what I did in #3502.

> Not all of them at once, every single.

You can discuss every single of the changes in the according PR. I'm not gonna open 100 single PRs.

> all DEBUG statements should be converted to LOG_DEBUG. Off by default. No change in policy. No radical change in code size for the default compiles. We can improve later.

That's what I do.

>  >Just let us make sure that everything that reveals a critical issue, gets printed with DEVELHELP.

> Don't you think that raising a log message's severity level is the wrong way? They're not even linked to DEVELHELP...

If would think so, I would not have proposed in the first place. What do you mean by "linked to DEVELHELP"?

> I think we should draft up and discuss the concept of "logging" in a more general way.

Yeah, we can do this, but since no one did it for the past two years, I think we ended up in a pretty ugly situation and every change can only make it better IMO.

**Comment by OlegHahm** [comment]
Btw. if you are worried about code size:
1. Build without `DEVELHELP`.
2. Decrease the log level.
3. Use `sys_log_printfnoformat`

**Comment by kaspar030** [comment]
>> all DEBUG statements should be converted to LOG_DEBUG. Off by default. No change in policy. No radical change in code size for the default compiles. We can improve later.
 
> That's what I do.

No. You replace most DEBUGs with LOG_DEBUG, some with LOG_WARNING, some with LOG_ERROR. Changing policy.

That is, IMHO, wrong. For many cases, it leads to a lot of useless .text.

For example, netdev_eth returns ENOENT or ENOTSUP as error code in some places, but additionally now has a raised-over-DEBUG log statement. That is log bloat, the error should be printed where it arrives, e.g., in the caller, unless specifically requested for that file, using either DEBUG_LEVEL>=LOG_DEBUG or ENABLE_DEBUG==1.

> What do you mean by "linked to DEVELHELP"?

Does DEVELHELP change the log level? (Ah, you made it change the level).

IMO when introducing the logging API, didn't we come to the conclusion that for most actual debugging needs, we'd keep DEBUG, as a global DEBUG flag would be way too verbose?

Logging should be used for anything the user might want to know about a working system, not for developers of a specific subsystem.

I'm fine with using LOG_DEBUG instead of DEBUG for unification's sake, but with a) replacing broken error handling, b) sacrificing our embedded code size requirements and c) practically removing non-debug log statements.

To cut this short, I'll not stop arguing until this PR only replaces DEBUG with LOG_DEBUG. :)

**Comment by OlegHahm** [comment]
> No. You replace most DEBUGs with LOG_DEBUG, some with LOG_WARNING, some with LOG_ERROR.

Actually, it was the other way around. I wanted to unveil some of the error messages that were hidden in debug mode. Might be that I overdid it. We can go through it one by one. I will start to overhaul it.

> IMO when introducing the logging API, didn't we come to the conclusion that for most actual debugging needs, we'd keep DEBUG, as a global DEBUG flag would be way too verbose?
>
> Logging should be used for anything the user might want to know about a working system, not for developers of a specific subsystem.
>
> I'm fine with using LOG_DEBUG instead of DEBUG for unification's sake,

That's why I asked and put these changes into a separate commit. I don't have a strong feeling either, but think that having `DEBUG` and `LOG_DEBUG` might be superfluous and confusing.

> a) replacing broken error handling

I don't get this. Error handling is left untouched. I agree that we might want to add some assertions.

> b) sacrificing our embedded code size requirements 

See my previous comment.

> c) practically removing non-debug log statements.

Aren't you arguing for the contrary?

> To cut this short, I'll not stop arguing until this PR only replaces DEBUG with LOG_DEBUG. :)

I would rather close this PR. The main purpose of this PR was to unveil hidden error messages. Changing `DEBUG` to `LOG_DEBUG` was more an unimportant side effect.

**Comment by kaspar030** [comment]
@OlegHahm Close?

**Comment by OlegHahm** [comment]
Yes, I can make a better proposal soon.

**Pull request closed without merging**